### PR TITLE
Tile hashing fix

### PR DESCRIPTION
**Pull request by phanecak-maptiler**
While generating maps for Slovakia and some other countries I've noticed that certain tiles are still stored in the .mbtiles more than once, e.g. with same encoded content, despite what was done in PR #219 .

For smaller maps (like for Slovakia), following can be used directly in SQLite to detect duplicates:

```
SELECT count(tile_data_id) as id_count, group_concat(tile_data_id) as id_list, substr(hex(tile_data), 0, 200) as tile_data_hex FROM tiles_data GROUP BY hex(tile_data) ORDER BY id_count DESC LIMIT 3;
```

result:

```
5|1267,1301,4515,16422,16423|1F8B08000000000000FF9312E1E2CDCD2FCD2B49CCCC8B2F484DCCD66850A860020067B86CD116000000
3|703,2324,7959|1F8B08000000000000FF03000000000000000000
1|19366|1F8B08000000000000FFEDBD696053D5D63FDC7D4E86739234C3C9D06427CDD434C3E994A94D9B4EA0800828C8E0AC58A042055A2CA057EFE3B594A922025644ACC85011111101111011102A42412E22222222222272111111918BF8DF43B8F6E4B9FFF
```

(Query is not very nice and does not scale well to bigger maps, but works.)

First observation was, that heuristic determining when to compute hashes was skipping those tiles. Then it was observed that hash calculation done in `FeatureFroup` (see https://github.com/onthegomap/planetiler/blob/3fd094d/planetiler-core/src/main/java/com/onthegomap/planetiler/collection/FeatureGroup.java#L377 ) is producing different hashes for tiles which when encoded and compressed, are byte-to-byte same.

Hence this PR does following:

1. `BikeRouteOverlayTest` unit test was extended to count duplicates in generated .mbtiles
    - I was not able to make some new reasonable unit test to cover this case, hence `BikeRouteOverlayTest` was identified as triggering the root-causes and thus extended to determine duplicate tiles.
2. heuristic determining hash calculation was extended to cover also empty tiles and tiles with just one empty layer
3. hash computation was adjusted to use result of layer encoding
   - Layer encoded to bytes is what is at the end written to .mbtiles. And I take it is easier and more error prone than trying to adjust original hash function which is working with `FeatureGroup` and features.

Updated Planetiler seems working better (=fewer duplicates, same performance), at least based on the unit test and a test with `slovakia.mbtiles`:

```
SELECT count(tile_data_id) as id_count, group_concat(tile_data_id) as id_list, substr(hex(tile_data), 0, 200) as tile_data_hex FROM tiles_data GROUP BY hex(tile_data) ORDER BY id_count DESC LIMIT 3;
1|19360|1F8B08000000000000FFEDBD69605455D6369A7D4E0DE754556A3835A46A57A5A6546A38996A4A2AA94CA0800828C8E0AC18204204120CA0ADFDDA8630478610634444848888880888808880808848D38848D3888888883422D2888869FCF650B439F5F
1|7|1F8B08000000000000FFEDBC799865659D26F8EDDBD9CF3D778FE5C6CDD83233724F9224933549927D914DCB298B099210824C223032130AAABB15524B4DA4B4AAC4056D05C442298534052BB5102C53EC6AA60BC12E4AF429A703ADD21A977AA6797AE
1|1736|1F8B08000000000000FFEDBC79985D55992FBCDFB5A7B5F63C4F679EA79A4ECD4392AA4A659EE790040827950AA9A45215AA1220D89AC810110101111051898D083882338D880C6AA3AD69B4FB3ACB756A1BED960FD1EEAB36F75BE724206D777FF77ED
```

(No count bigger than 1, hence no duplicates.)

Side note reg. `BikeRouteOverlayTest`: While debugging, following was observed:

```
[mbtiles:encode] old hash: 8152459079797329014, new hash: -7108558662289803811
[mbtiles:encode] old hash: 8136981643901169510, new hash: -7108558662289803811
[mbtiles:encode] old hash: -4610225015425806810, new hash: -7108558662289803811
[mbtiles:encode] old hash: 8561865035814504758, new hash: -7108558662289803811
[mbtiles:encode] old hash: -7291151759222358010, new hash: -7108558662289803811
[mbtiles:encode] old hash: -8581161042698348874, new hash: -7108558662289803811
```

### CODE REVIEW

**Comment by msbarry** [comment]
I'm not quite sure why the actions didn't run so I merged in main and pushed - they seem to be going now.

File: planetiler-examples/src/test/java/com/onthegomap/planetiler/examples/BikeRouteOverlayTest.java [hunk 1]
```diff
@@ -110,6 +110,8 @@ void integrationTest(@TempDir Path tmpDir) throws Exception {
           "name", "EuroVelo 8 - Mediterranean Route - part Monaco",
           "ref", "EV8"
         ), GeoUtils.WORLD_LAT_LON_BOUNDS, 25, LineString.class);
+
+      TestUtils.assertTileDuplicates(mbtiles, 0);
```
Comment (msbarry) [inline planetiler-examples/src/test/java/com/onthegomap/planetiler/examples/BikeRouteOverlayTest.java:114]: Is there somewhere in the core module we can put this? I just want to make sure core functionality is verified there in case we remove/change examples in the future.

Comment (phanecak-maptiler) [reply planetiler-examples/src/test/java/com/onthegomap/planetiler/examples/BikeRouteOverlayTest.java:114]: So far no other UT is triggering the root-cause.

Hence I've tried to make a new UT in `planetiler-core` but then realized I would need to replicate quite a lot of code from `planetiler-examples`.

E.g. both options are "bad".

Conclusion so far was to go with "smaller bad", e.g. the current option, since it adds much smaller amount of code.

File: planetiler-core/src/test/java/com/onthegomap/planetiler/collection/FeatureGroupTest.java [hunk 2]
```diff
@@ -356,17 +359,22 @@ void testHasSameContents(String testName, boolean expectSame, PuTileArgs args0,
...
-  void testGenerateContentHash(String testName, boolean expectSame, PuTileArgs args0, PuTileArgs args1) {
+  void testGenerateContentHash(String testName, boolean expectSame, PuTileArgs args0, PuTileArgs args1)
+    throws IOException {
     put(args0);
     put(args1);
     sorter.sort();
     var iter = features.iterator();
-    var tile0 = iter.next();
-    var tile1 = iter.next();
+    var tile0_hash = MbtilesWriter.generateContentHash(
```
[the remark this answers is not in the document]
Comment (phanecak-maptiler) [reply planetiler-core/src/test/java/com/onthegomap/planetiler/collection/FeatureGroupTest.java:368]: Yup, that's what you get when switching between Java and Python too often. :) Sorry for that.

File: planetiler-core/src/main/java/com/onthegomap/planetiler/mbtiles/MbtilesWriter.java [hunk 3]
```diff
@@ -412,6 +413,17 @@ private long tilesEmitted() {
...
+   * Generates a hash over encoded and compressed tile.
+   * <p>
+   * Used as an optimization to avoid writing the same (mostly ocean) tiles over and over again.
+   */
+  public static long generateContentHash(byte[] bytes) {
+    long hash = Hashing.FNV1_64_INIT;
+    hash = Hashing.fnv1a64(hash, bytes);
+    return hash;
+  }
+
```
[the remark this answers is not in the document]
Comment (phanecak-maptiler) [reply planetiler-core/src/main/java/com/onthegomap/planetiler/mbtiles/MbtilesWriter.java]: adjusted

File: planetiler-core/src/main/java/com/onthegomap/planetiler/VectorTile.java [hunk 4]
```diff
@@ -523,6 +523,19 @@ private boolean containsOnlyFillsOrEdges(boolean allowEdges) {
     return !empty;
   }
 
+  public boolean likelyToBeDuplicate() {
```
[the remark this answers is not in the document]
Comment (phanecak-maptiler) [reply planetiler-core/src/main/java/com/onthegomap/planetiler/VectorTile.java]: Adjusted and javadoc added. I've reused stuff you've mentioned on Slack. Thanks for the feedback.

File: planetiler-core/src/main/java/com/onthegomap/planetiler/mbtiles/MbtilesWriter.java [hunk 5]
```diff
@@ -289,8 +290,8 @@ private void tileEncoder(Iterable<TileBatch> prev, Consumer<TileBatch> next) thr
           lastEncoded = encoded;
           lastBytes = bytes;
           last = tileFeatures;
-          if (compactDb && en.containsOnlyFillsOrEdges()) {
-            tileDataHash = tileFeatures.generateContentHash();
+          if (compactDb && en.likelyToBeDuplicate() && bytes != null) {
+            tileDataHash = generateContentHash(bytes);
```
[the remark this answers is not in the document]
Comment (phanecak-maptiler) [reply planetiler-core/src/main/java/com/onthegomap/planetiler/mbtiles/MbtilesWriter.java:294]: From top of the head I would say that while different implementations may produce different results for same input, I believe same implementation should always produce same results for same input, unless we're tweaking some parameters (like compression level) between the runs.

I did not find a definitive mathematical proof for that, but quick skim through OpenJDK sources and following references confirm the assumption "same input -> same output":

- https://laptrinhx.com/is-gzip-deterministic-4247081585/
- https://dramsch.net/today-i-learned/gzip/today-i-learned-about-deterministic-gzip-compression/

TL;DR: Output is deterministic, unless we're compressing several files which do have same content but have different filesystem metadata.

**Comment by phanecak-maptiler** [comment]
As per "Example project" failure: It looks like chicken-egg problem: unless we merge this, it will fail, since it is running with `planetiler-core` produced without this PR.

**Comment by msbarry** [comment]
Looks good! Thanks for the fix.
